File: android/app/src/main/java/com/deviceguard/core/permissions/PermissionGuide.kt
package com.deviceguard.core.permissions

import com.deviceguard.core.guardengine.DeviceState

enum class GuardPermissionKind {
    BLUETOOTH,
    NOTIFICATION,
    LOCAL_NETWORK,
    FOREGROUND_SERVICE,
    BATTERY_OPTIMIZATION
}

enum class GuardPermissionStatus {
    GRANTED,
    DENIED,
    NOT_DETERMINED,
    RESTRICTED
}

data class GuardPermissionSnapshot(
    val bluetooth: GuardPermissionStatus,
    val notification: GuardPermissionStatus,
    val localNetwork: GuardPermissionStatus,
    val foregroundService: GuardPermissionStatus,
    val batteryOptimization: GuardPermissionStatus
)

data class PermissionIssue(
    val kind: GuardPermissionKind,
    val title: String,
    val message: String,
    val primaryActionTitle: String,
    val degradedState: DeviceState?
)

class PermissionGuide {
    fun mostImportantIssue(snapshot: GuardPermissionSnapshot): PermissionIssue? {
        if (snapshot.bluetooth != GuardPermissionStatus.GRANTED) {
            return PermissionIssue(
                kind = GuardPermissionKind.BLUETOOTH,
                title = "需要开启附近设备权限",
                message = "用于发现看护组内手机是否仍在身边。",
                primaryActionTitle = "去开启",
                degradedState = DeviceState.PERMISSION_LIMITED
            )
        }

        if (snapshot.notification != GuardPermissionStatus.GRANTED) {
            return PermissionIssue(
                kind = GuardPermissionKind.NOTIFICATION,
                title = "需要开启通知",
                message = "设备可能离开时，通知用于及时提醒你。",
                primaryActionTitle = "去开启",
                degradedState = null
            )
        }

        if (snapshot.foregroundService != GuardPermissionStatus.GRANTED) {
            return PermissionIssue(
                kind = GuardPermissionKind.FOREGROUND_SERVICE,
                title = "后台看护可能受限",
                message = "开启持续看护时，系统需要显示看护状态通知。",
                primaryActionTitle = "查看设置",
                degradedState = null
            )
        }

        if (snapshot.localNetwork != GuardPermissionStatus.GRANTED) {
            return PermissionIssue(
                kind = GuardPermissionKind.LOCAL_NETWORK,
                title = "建议允许本地网络",
                message = "同一 Wi-Fi 下可减少室内误报；不开启时仍可使用蓝牙看护。",
                primaryActionTitle = "去开启",
                degradedState = null
            )
        }

        if (snapshot.batteryOptimization != GuardPermissionStatus.GRANTED) {
            return PermissionIssue(
                kind = GuardPermissionKind.BATTERY_OPTIMIZATION,
                title = "后台看护可能被系统限制",
                message = "部分手机会限制后台运行，可在系统设置中允许持续看护。",
                primaryActionTitle = "查看设置",
                degradedState = null
            )
        }

        return null
    }

    fun canStartCoreGuarding(snapshot: GuardPermissionSnapshot): Boolean {
        return snapshot.bluetooth == GuardPermissionStatus.GRANTED
    }
}

File: ios/GuardApp/Core/Permissions/PermissionGuide.swift
import Foundation

enum GuardPermissionKind: String, Equatable, CaseIterable {
    case bluetooth
    case notification
    case localNetwork
    case backgroundRefresh
}

enum GuardPermissionStatus: String, Equatable {
    case granted
    case denied
    case notDetermined
    case restricted
}

struct GuardPermissionSnapshot: Equatable {
    var bluetooth: GuardPermissionStatus
    var notification: GuardPermissionStatus
    var localNetwork: GuardPermissionStatus
    var backgroundRefresh: GuardPermissionStatus
}

struct PermissionIssue: Equatable {
    let kind: GuardPermissionKind
    let title: String
    let message: String
    let primaryActionTitle: String
    let degradedState: DeviceState?
}

final class PermissionGuide {
    func mostImportantIssue(from snapshot: GuardPermissionSnapshot) -> PermissionIssue? {
        if snapshot.bluetooth != .granted {
            return PermissionIssue(
                kind: .bluetooth,
                title: "需要开启蓝牙",
                message: "用于发现看护组内手机是否仍在身边。",
                primaryActionTitle: "去开启",
                degradedState: .permissionLimited
            )
        }

        if snapshot.notification != .granted {
            return PermissionIssue(
                kind: .notification,
                title: "需要开启通知",
                message: "设备可能离开时，通知用于及时提醒你。",
                primaryActionTitle: "去开启",
                degradedState: nil
            )
        }

        if snapshot.localNetwork != .granted {
            return PermissionIssue(
                kind: .localNetwork,
                title: "建议开启本地网络",
                message: "同一 Wi-Fi 下可减少室内误报；不开启时仍可使用蓝牙看护。",
                primaryActionTitle: "去开启",
                degradedState: nil
            )
        }

        if snapshot.backgroundRefresh != .granted {
            return PermissionIssue(
                kind: .backgroundRefresh,
                title: "后台看护可能受限",
                message: "系统限制后台刷新时，回到 App 后会立即更新状态。",
                primaryActionTitle: "查看设置",
                degradedState: nil
            )
        }

        return nil
    }

    func canStartCoreGuarding(snapshot: GuardPermissionSnapshot) -> Bool {
        snapshot.bluetooth == .granted
    }
}
